infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/StateTransferBatchProcessor.java

```diff
@@ -43,54 +48,68 @@
+                lastWriteException = Optional.of(e);
+                Sleep.sleepUninterruptibly(writeSleepDuration);
+                // Get all the addresses that were supposed to be written to a stream log.
+                long start = totalAddressesInRequest.get(0);
+                long end = totalAddressesInRequest.get(totalAddressesInRequest.size() - 1);
+                Set<Long> knownAddresses = CFUtils
+                        .getUninterruptibly(logUnitClient.requestKnownAddresses(start, end))
+                        .getKnownAddresses();
+                // Get all the addresses that were not written.
+                Set<Long> nonWrittenAddresses =
+                        Sets.difference(new HashSet<>(totalAddressesInRequest), knownAddresses);
+                // If the delta is empty -> return with success.
+                if (nonWrittenAddresses.isEmpty()) {
                     transferBatchResponse = Optional.of(TransferBatchResponse
                             .builder()
-                            .transferBatchRequest(new TransferBatchRequest(totalAddressesInRequest, destination))
+                            .transferBatchRequest(new TransferBatchRequest(totalAddressesInRequest,
+                                    readBatch.getDestination()))
                             .status(SUCCEEDED)
                             .build());
                     break;
                 }
-            } catch (Exception e) {
-                lastWriteException = Optional.of(e);
-                Sleep.sleepUninterruptibly(writeSleepDuration);
-                List<Long> remainingAddressesToWrite = readBatch.getAddresses();
-                // Get all the addresses that were supposed to be written to a stream log.
-                long start = remainingAddressesToWrite.get(0);
-                long end = remainingAddressesToWrite.get(remainingAddressesToWrite.size() - 1);
-                Set<Long> knownAddresses =
-                        logUnitClient.requestKnownAddresses(start, end).join().getKnownAddresses();
-                // Get all the addresses that were not written.
-                Set<Long> nonWrittenAddresses =
-                        Sets.difference(new HashSet<>(remainingAddressesToWrite), knownAddresses);
                 // Create a new read batch with the missing data and retry.
                 ImmutableList<LogData> nonWrittenData = readBatch.getData()
                         .stream()
                         .filter(data -> nonWrittenAddresses.contains(data.getGlobalAddress()))
                         .collect(ImmutableList.toImmutableList());
                 readBatch = readBatch.toBuilder().data(nonWrittenData).build();
+            } catch (RuntimeException re) {
+                // If it's some other exception, e.g. IllegalArgumentException,
+                // WrongEpochException -> Propagate it to the caller.
+                throw new IllegalStateException(re);
```

Comment:
this seems like a mess, would be great if you can come up with a better solution but I understand if the current approach makes smaller impact on the system